### ISSUE TITLE
Drop basicAuth on external schemas..?

### ISSUE THREAD

**Issue by xeor** [mvantellingen/python-zeep#588]
1. The version of zeep (or if you are running master the commit hash/date): `2.4.0`
2. The WSDL you are using
It is very complex, but the beginning (relevant part) is:
```
<definitions xmlns="http://schemas.xmlsoap.org/wsdl/"
   xmlns:cmn="http://schemas.hp.com/SM/7/Common"
   xmlns:http="http://schemas.xmlsoap.org/wsdl/http/"
   xmlns:mime="http://schemas.xmlsoap.org/wsdl/mime/"
   xmlns:ns="http://schemas.hp.com/SM/7"
   xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/"
   xmlns:xsd="http://www.w3.org/2001/XMLSchema"
   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
   targetNamespace="http://schemas.hp.com/SM/7" 
   xsi:schemaLocation="http://schemas.xmlsoap.org/wsdl/ http://schemas.xmlsoap.org/wsdl/">
<types>
<xs:schema
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  xmlns="http://schemas.hp.com/SM/7"
  xmlns:cmn="http://schemas.hp.com/SM/7/Common"
  xmlns:xmime="http://www.w3.org/2005/05/xmlmime"
  attributeFormDefault="unqualified"
  elementFormDefault="qualified"
  targetNamespace="http://schemas.hp.com/SM/7"
  version="2017-06-15 Rev 2">

<xs:import
  namespace="http://www.w3.org/2005/05/xmlmime" 
  schemaLocation="http://www.w3.org/2005/05/xmlmime"/>

.... lots of non-relevant xml dropped ...
```

3. An example script
Basicly the same as http://docs.python-zeep.org/en/master/transport.html#http-authentication
```py
from requests import Session
from requests.auth import HTTPBasicAuth
from zeep import Client
from zeep.transports import Transport

session = Session()
session.auth = HTTPBasicAuth(user, password)
client = Client('http://my-endpoint.com/production.svc?wsdl',
    transport=Transport(session=session))
```

The problem is that the external schema in this case (`http://www.w3.org/2005/05/xmlmime`), contains the header for Basic auth (`Authorization: Basic abc123`). Ie not only is this sent to `http://my-endpoint.com/production.svc?wsdl` but also 3rd party urls...

Is this by design? Any way around it? Is it a part of the soap-spec? In my eyes, it is a security issue waiting to bite!

**Comment by xeor** [comment]
A simple fix would be for the client.transport to support using a dict instead.
Or maybe a better idea would be to extend the `plugins` support to support something like `get_transport(location)`..?

**Comment by filipposantovito** [comment]
just an idea:
maybe you can use [custom-authentication](http://docs.python-requests.org/en/master/user/advanced/#custom-authentication) and filter out your credentials when not calling the right domain.

**Comment by xeor** [comment]
How did I miss `Custom Authentication`? :laughing: 

I made a simple custom authentication for my use-case, might be what someone else needs as well, so here it is:

```py
import re
from requests import auth

class BasicAuthPerDomain(auth.AuthBase):
    def __init__(self, matchmap):
        self.matchmap = matchmap

    def __call__(self, r):
        # matcher will be the last 2 domains in the url.
        # something.example.com:123/blah will be "example.com"
        matcher = '.'.join(re.findall(r'^https?://([^/:]+)', r.url)[0].split('.')[-2:])

        matched_auth = self.matchmap.get(matcher)
        if auth:
            r.headers['Authorization'] = auth._basic_auth_str(*matched_auth)
        return r
```

Use it like so:

```py
session.auth = BasicAuthPerDomain({
    'example.no': ('username', 'password'),
    'another-example.no': ('u1', 'p2'),
})
```

I still think the original issue is weird tho... Should it be like this?

**Comment by filipposantovito** [comment]
I think this is a requests problem and, if you want, you can submit a PR to https://github.com/requests/requests . 

what if you set credentials for a session object and then the http server redirects your request to another domain asking for http basic auth? I suppose the object will happily transmit your credentials to the new site. Maybe this is not the right behaviour, maybe yes. Probably it should be a configurable behaviour.

**Comment by xeor** [comment]
The session is one per `Transport` and it is reused on every requests.. So it looks like it will happily share cookies, auth-tokens and everything else request sessions stores.

https://github.com/mvantellingen/python-zeep/blob/2.4.0/src/zeep/wsdl/wsdl.py#L139-L140
ends up requesting using the session @ https://github.com/mvantellingen/python-zeep/blob/2.4.0/src/zeep/transports.py#L125

**Comment by mvantellingen** [comment]
Yeah this seems to be a problem. Perhaps we need to stop re-using the same session and just create a new session for each request? And perhaps only add the auth credentials on the initial url?

If anyone wants to help me create a PR for this then that would be really appreciated.
